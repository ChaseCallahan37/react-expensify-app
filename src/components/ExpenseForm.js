import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    const { expense } = props;
    this.state = {
      description: expense ? expense.description : "",
      note: expense ? expense.note : "",
      amount: expense ? (expense.amount / 100).toString() : "",
      createdAt: expense ? moment(expense.createdAt) : moment(),
      calFocus: false,
      error: "",
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState({ description });
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState({ note });
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount });
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState({ createdAt });
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState({ calFocus: focused });
  };
  onSubmit = (e) => {
    const { description, amount, error, createdAt, note } = { ...this.state };
    e.preventDefault();
    if (!description || !amount) {
      return this.setState({
        error: "Please provide a description and amount!",
      });
    }
    this.setState({ error: "" });

    this.props.onSubmit({
      description,
      amount: parseFloat(amount, 10) * 100,
      createdAt: createdAt.valueOf(),
      note,
    });
  };
  render() {
    const { description, note, amount, createdAt, calFocus, error } =
      this.state;
    return (
      <div>
        {error && <h2>{error}</h2>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={description}
            onChange={this.onDescriptionChange}
          />
          <SingleDatePicker
            date={createdAt}
            onDateChange={this.onDateChange}
            focused={calFocus}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input
            type="text"
            placeholder="Amount"
            value={amount}
            onChange={this.onAmountChange}
          />
          <textarea
            placeholder="Add a note"
            value={note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
