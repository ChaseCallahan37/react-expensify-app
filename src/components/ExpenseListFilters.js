import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    debugger;
    this.setState({ calendarFocused });
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };
  render() {
    const { filters, dispatch } = this.props;
    const { calendarFocused } = this.state;
    return (
      <div>
        <input type="text" value={filters.text} onChange={this.onTextChange} />
        <select onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate}
          endDate={filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          showClearDates={true}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  endStartDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
