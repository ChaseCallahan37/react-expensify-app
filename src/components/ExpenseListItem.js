import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";
import { NavLink } from "react-router-dom";

export const ExpenseListItem = ({
  id,
  description,
  note,
  amount,
  createdAt,
  dispatch,
}) => {
  return (
    <li>
      <NavLink to={`/edit/${id}`}>
        <h3>{description} </h3>
      </NavLink>
      <label>
        {` $${amount} `}
        {`Created at: ${createdAt}`}
      </label>
      <br></br>
      <label>{note}</label>
      <button
        onClick={(e) => {
          dispatch(removeExpense(id));
        }}
      >
        X
      </button>
    </li>
  );
};

export default connect()(ExpenseListItem);
