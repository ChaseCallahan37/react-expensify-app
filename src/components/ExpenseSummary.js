import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import getExpenseTotal from "../selectors/expense-total";
import expenseSelector from "../selectors/expenses";

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  const total = numeral(expenseTotal).format("$0,0.00");
  const expenseWording = expenseCount.length === 1 ? "expense" : "expenses";
  return (
    <div>
      <h1>
        There are {expenseCount} {expenseWording} totaling: {total}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const filteredExp = expenseSelector(state.expenses, state.filters);
  return {
    expenseCount: filteredExp.length,
    expenseTotal: getExpenseTotal(filteredExp) / 100,
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
