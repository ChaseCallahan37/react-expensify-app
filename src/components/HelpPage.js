import React from "react";

const HelpPage = () => <div>this is the help expense Dashboard</div>;

export default HelpPage;
