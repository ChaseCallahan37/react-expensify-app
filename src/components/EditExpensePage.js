import React from "react";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(expense.id, expense);
    this.props.history.push("/");
  };
  removeExpense = () => {
    const { removeExpense, expense } = this.props;
    removeExpense(expense.id);
    this.props.history.push("/");
  };
  render() {
    const { expense, match } = this.props;
    return (
      <div>
        <ExpenseForm expense={expense} onSubmit={this.onSubmit} />
        <button type="button" onClick={this.removeExpense}>
          Remove Expense
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((e) => e.id === props.match.params.id),
  };
};

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
