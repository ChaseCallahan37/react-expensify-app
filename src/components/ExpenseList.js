import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectedExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
  const { expenses } = props;
  return (
    <div>
      <h1>Expense List</h1>
      {expenses.length === 0 ? (
        <p>There are no Expenses</p>
      ) : (
        <ul>
          {expenses.map((e) => {
            return <ExpenseListItem key={e.id} {...e} />;
          })}
        </ul>
      )}
    </div>
  );
};

const mapStateToProps = ({ expenses, filters }) => {
  return {
    expenses: selectedExpenses(expenses, filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
