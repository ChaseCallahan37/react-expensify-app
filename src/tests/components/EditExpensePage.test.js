import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { EditExpensePage } from "../../components/EditExpensePage";

let wrapper, editExpense, removeExpense, match, history, id;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});

test("Should render EditExpensePage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should properly handle edit expense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("Should handle remove expense function", () => {
  wrapper.find("button").prop("onClick")();
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[1].id);
  expect(history.push).toHaveBeenLastCalledWith("/");
});
