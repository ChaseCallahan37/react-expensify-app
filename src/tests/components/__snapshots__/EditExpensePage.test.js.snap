// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 200,
        "createdAt": 345600000,
        "description": "Gum",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="button"
  >
    Remove Expense
  </button>
</div>
`;
